src/core/queryClient.ts

```diff
@@ -532,13 +533,35 @@ export class QueryClient {
     }
   }
 
+  findFirstMatchingQueryDefaults(
+    queryKey?: QueryKey
+  ): QueryOptions<any, any, any> | undefined {
+    if (!queryKey) {
+      return undefined
+    }
+
+    // First retrieve all matching defaults for the given key
+    const matchingDefaults = this.queryDefaults.filter(x =>
+      partialMatchKey(queryKey, x.queryKey)
+    )
+    // It is ok not having defaults, but it is error prone to have more than 1 default for a given key
+    if (matchingDefaults?.length > 1) {
+      getLogger().warn(
+        `[QueryClient] Several defaults match with key '${JSON.stringify(
+          queryKey
+        )}'. The first matching query defaults are used. Please check how query defaults are registered. Order does matter here. cf. https://react-query.tanstack.com/reference/QueryClient#queryclientsetquerydefaults.`
+      )
+    }
+    // Explicitly returns the first one
+    const firstMatchingDefaults = matchingDefaults?.[0]
+    return firstMatchingDefaults?.defaultOptions || undefined
+  }
+
   getQueryDefaults(
```

Comment:
what about `getMutationDefaults` ?

Reply:
I haven't played much with mutation yet. Let me check and apply the same pattern if needed.

Reply:
Indeed, both query and mutation defaults are working the same way.
I'm not familiar enough with TS to find a way to factorise the code, so I duplicated the code and changed what was needed.

I still think factorisation is required here. WDYT?
Could you give me some hints on how both `findFirstMatchingQueryDefaults` and `findFirstMatchingMutationDefaults` could be merged? `findFirstMatchingDefaults(defaultsStore, queryKey)` could do it. But I'm struggling with overload…

Reply:
the main goal should be:
- not affect production bundle size
- not affect production performance

right now, we `filter` even though a `find` would be sufficient, which can terminate early. Also, we add a bit of code for production that is not necessary.

How about:
- doing the `.find` like before
- adding an env check, behind which we can:
  - do an additional `filter`
  - log if that produces more than one result

that code could also stay duplicated because it is stripped way in production anyhow. And during dev mode, it doesn't matter if we do find + filter, as in prod, it will stay only one `find`

Reply:
Yep will do as suggested. I don't think the number of defaults would be too high but I agree, let's keep production code as performant as it could be.